test/diff_model.jl

```diff
@@ -93,3 +93,44 @@ end
 
     @test maximum(abs.(grad_wrt_h - [0.25; 0.75; -1.75])) <= 1e-4
 end
+
+
+@testset "Differentiating a non-convex QP" begin
+    Q = [0.0 0.0; 1.0 2.0]
+    q = [1.0; 1.0]
+    G = [1.0 1.0;]
+    h = [-1.0;]
+
+    model = MOI.instantiate(OSQP.Optimizer, with_bridge_type=Float64)
+    x = MOI.add_variables(model, 2)
+
+    # define objective
+    quad_terms = MOI.ScalarQuadraticTerm{Float64}[]
+    for i in 1:2
+        for j in i:2 # indexes (i,j), (j,i) will be mirrored. specify only one kind
+            push!(
+                quad_terms, 
+                MOI.ScalarQuadraticTerm(Q[i,j], x[i], x[j])
+            )
+        end
+    end
+
+    objective_function = MOI.ScalarQuadraticFunction(
+                            MOI.ScalarAffineTerm.(q, x),
+                            quad_terms,
+                            0.0
+                        )
+    MOI.set(model, MOI.ObjectiveFunction{MOI.ScalarQuadraticFunction{Float64}}(), objective_function)
+    MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)
+
+    # add constraint
+    MOI.add_constraint(
+        model,
+        MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(G[1, :], x), 0.0),
+        MOI.LessThan(h[1])
+    )
+
+    diff = diff_model(model)
+
+    @test_throws ErrorException diff.forward() # should break
```

Comment:
ideally, we should use a more specific exception instead of Error, so defining a custom:
`struct NonConvexException <: Exception end`
and then `throw(NonConvexException())` instead of error(something)

Reply:
ok. i agree. so should we check for convexity ourselves?

Reply:
in the current implementation, which part is throwing the error?

Reply:
its actually `OSQP.jl`. The error is invoked when we call `optimize!` in the `forward` method

Reply:
then I don't think we should test the error on non-convexity, since it creates a test that works on one solver but not or others.